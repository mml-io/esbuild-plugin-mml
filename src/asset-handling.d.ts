import * as esbuild from "esbuild";
import { Asset } from "./types";
export declare function createAssetHandler(sourceRoot: string, outdir: string, assetDir: string, assetPrefix: string, assets: Asset[], build: esbuild.PluginBuild): (args: esbuild.OnLoadArgs) => Promise<esbuild.OnLoadResult | null>;
