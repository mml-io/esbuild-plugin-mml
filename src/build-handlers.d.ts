import * as esbuild from "esbuild";
import { DiscoveredEntryPoints } from "./types";
export declare function createEntryPointResolver(discoveredEntryPoints: DiscoveredEntryPoints, build: esbuild.PluginBuild): (args: esbuild.OnResolveArgs) => Promise<esbuild.OnResolveResult | null>;
export declare function createMMLImportResolver(build: esbuild.PluginBuild): (args: esbuild.OnResolveArgs) => Promise<esbuild.OnResolveResult>;
export declare function createMMLRefHandler(sourceRoot: string, entryPointToOutputUrl: Record<string, string>, documentPrefix: string): (args: esbuild.OnLoadArgs) => esbuild.OnLoadResult;
export declare function createMMLTSXRefHandler(sourceRoot: string, entryPointToOutputUrl: Record<string, string>, documentPrefix: string): (args: esbuild.OnLoadArgs) => esbuild.OnLoadResult;
export declare function createTSXHandler(): (args: esbuild.OnLoadArgs) => Promise<esbuild.OnLoadResult>;
export declare function createWorldHandler(): (args: esbuild.OnLoadArgs) => Promise<esbuild.OnLoadResult>;
export declare function createHTMLHandler(): (args: esbuild.OnLoadArgs) => Promise<esbuild.OnLoadResult>;
