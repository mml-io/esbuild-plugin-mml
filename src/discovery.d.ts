import * as esbuild from "esbuild";
import { DiscoveredEntryPoints } from "./types";
export declare function doDiscoveryBuild(esbuildOptions: esbuild.BuildOptions): Promise<DiscoveredEntryPoints>;
