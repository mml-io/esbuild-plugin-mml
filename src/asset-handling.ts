import * as esbuild from "esbuild";
import * as path from "path";
import fsp from "node:fs/promises";
import { Asset } from "./types";

export function createAssetHandler(
  sourceRoot: string,
  outdir: string,
  assetDir: string,
  assetPrefix: string,
  assets: Asset[],
  build: esbuild.PluginBuild,
  globalNamePrefix = "",
) {
  return async (
    args: esbuild.OnLoadArgs,
  ): Promise<esbuild.OnLoadResult | null> => {
    const extension = path.extname(args.path);

    // Check if this file should be handled as an asset
    const loader = build.initialOptions.loader?.[extension];
    if (loader !== "file") {
      return null;
    }

    // Generate asset file name based on the source path
    const relativePath = path.relative(sourceRoot, args.path);
    let fileName = relativePath.replace(/[/\\]/g, "_");

    // Apply globalNamePrefix to the asset file name if provided
    if (globalNamePrefix) {
      fileName = globalNamePrefix + "-" + fileName;
    }

    // Build full asset URL
    const assetUrl = assetPrefix + fileName;

    // Determine output path
    const outputPath = assetDir
      ? path.join(outdir, assetDir, fileName)
      : path.join(outdir, fileName);

    // Copy the asset to the output directory
    await fsp.mkdir(path.dirname(outputPath), { recursive: true });
    await fsp.copyFile(args.path, outputPath);

    // Track the asset
    assets.push({
      fileName,
      url: assetUrl,
      path: outputPath,
    });

    return {
      contents: assetUrl,
      loader: "text",
    };
  };
}
