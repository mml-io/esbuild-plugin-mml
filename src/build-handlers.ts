import * as esbuild from "esbuild";
import * as path from "path";
import fsp from "node:fs/promises";
import { DiscoveredEntryPoints } from "./types";

export function createEntryPointResolver(
  discoveredEntryPoints: DiscoveredEntryPoints,
  build: esbuild.PluginBuild,
) {
  return async (
    args: esbuild.OnResolveArgs,
  ): Promise<esbuild.OnResolveResult | null> => {
    if (args.kind !== "entry-point" || args.pluginData === "mml-skip-resolve") {
      return null;
    }

    const resolvedPath = await build.resolve(args.path, {
      resolveDir: args.resolveDir,
      kind: args.kind,
      pluginData: "mml-skip-resolve",
    });

    const namespace = discoveredEntryPoints.tsx.has(resolvedPath.path)
      ? "mml-build-tsx"
      : discoveredEntryPoints.html.has(resolvedPath.path)
        ? "mml-build-html"
        : discoveredEntryPoints.world.has(resolvedPath.path)
          ? "mml-build-world"
          : null;

    return namespace
      ? {
          path: resolvedPath.path,
          namespace,
          pluginData: resolvedPath.pluginData as unknown,
        }
      : null;
  };
}

export function createMMLImportResolver(build: esbuild.PluginBuild) {
  return async (
    args: esbuild.OnResolveArgs,
  ): Promise<esbuild.OnResolveResult> => {
    const importPath = args.path.replace(/^mml:/, "");
    const resolvedPath = await build.resolve(importPath, {
      resolveDir: args.resolveDir,
      kind: args.kind,
      pluginData: "mml-skip-resolve",
    });

    if (resolvedPath.errors.length > 0) {
      return {
        errors: resolvedPath.errors.map((error) => ({
          ...error,
          text: `Failed to resolve ${importPath} from ${args.importer}. ${error.text}`,
        })),
      };
    }

    return {
      path: resolvedPath.path,
      namespace: resolvedPath.path.endsWith(".html")
        ? "mml-build-ref"
        : "mml-build-tsx-ref",
      pluginData: resolvedPath.pluginData as string,
    };
  };
}

function createRefHandler(
  sourceRoot: string,
  entryPointToOutputUrl: Record<string, string>,
  documentPrefix: string,
  isTSX = false,
) {
  return (args: esbuild.OnLoadArgs): esbuild.OnLoadResult => {
    const relativePath = path.relative(sourceRoot, args.path);
    let outputUrl = entryPointToOutputUrl[relativePath];

    if (!outputUrl) {
      const baseName = path.basename(relativePath, path.extname(relativePath));
      const dirPath = path.dirname(relativePath);
      const pathParts = [dirPath, baseName].filter(
        (part) => part && part !== ".",
      );

      outputUrl =
        pathParts
          .join("-")
          .replace(isTSX ? /[^a-zA-Z0-9\-_]/g : /[/\\]/g, "-") + ".html";

      if (!isTSX) {
        entryPointToOutputUrl[relativePath] = outputUrl;
      }
    }

    return {
      contents: documentPrefix + outputUrl,
      loader: "text",
      resolveDir: path.dirname(args.path),
      watchFiles: [args.path],
    };
  };
}

export function createMMLRefHandler(
  sourceRoot: string,
  entryPointToOutputUrl: Record<string, string>,
  documentPrefix: string,
) {
  return createRefHandler(
    sourceRoot,
    entryPointToOutputUrl,
    documentPrefix,
    false,
  );
}

export function createMMLTSXRefHandler(
  sourceRoot: string,
  entryPointToOutputUrl: Record<string, string>,
  documentPrefix: string,
) {
  return createRefHandler(
    sourceRoot,
    entryPointToOutputUrl,
    documentPrefix,
    true,
  );
}

async function loadFileContent(
  args: esbuild.OnLoadArgs,
  loader: "tsx" | "text",
) {
  const content = await fsp.readFile(args.path, "utf-8");
  return {
    contents: content,
    loader,
    resolveDir: path.dirname(args.path),
    watchFiles: [args.path],
  };
}

export const createTSXHandler = () => (args: esbuild.OnLoadArgs) =>
  loadFileContent(args, "tsx");

export const createWorldHandler = () => (args: esbuild.OnLoadArgs) =>
  loadFileContent(args, "tsx");

export const createHTMLHandler = () => (args: esbuild.OnLoadArgs) =>
  loadFileContent(args, "text");
