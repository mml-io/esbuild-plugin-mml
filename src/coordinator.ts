import * as esbuild from "esbuild";
import { makeResultProcessor } from "./results";
import { worldContext } from "./world";
import { documentContext } from "./documents";

export interface CoordinatorOptions {
  verbose: boolean;
  options: esbuild.BuildOptions;
  worlds: string[];
  documents: string[];
  assetDir: string;
  processor: ReturnType<typeof makeResultProcessor>;
  build: esbuild.PluginBuild["esbuild"];
}

function subsetOf<T>(set: Set<T>, subset: Set<T>): boolean {
  return Array.from(subset).every((v) => set.has(v));
}

function union<T>(s1: Set<T>, s2: Set<T>): Set<T> {
  const result = new Set<T>();
  s1.forEach((v) => result.add(v));
  s2.forEach((v) => result.add(v));
  return result;
}

export class Coordinator {
  private running = false;
  private building = true;
  private documents: Set<string>;
  private discoveredDocuments = new Set<string>();
  private docsCtx?: esbuild.BuildContext;
  private worldCtx?: esbuild.BuildContext;

  constructor(private options: CoordinatorOptions) {
    this.documents = new Set(options.documents);
  }

  public async start() {
    this.worldCtx = await worldContext({
      ...this.options,
      onEnd: (result, discoveredDocuments, importStubs) => {
        discoveredDocuments.forEach((d) => this.discoveredDocuments.add(d));
        this.options.processor.pushResult("world", result, importStubs);
      },
    });

    void this.worldCtx.watch();

    this.running = true;
    await this.tick(true);
  }

  async tick(firstTick = false) {
    if (firstTick || !subsetOf(this.documents, this.discoveredDocuments)) {
      this.documents = union(this.documents, this.discoveredDocuments);
      this.discoveredDocuments.clear();

      const oldDocsCtx = this.docsCtx;

      this.docsCtx = await documentContext({
        ...this.options,
        documents: Array.from(this.documents),
        onDiscoveredDocuments: (discoveredDocuments) => {
          discoveredDocuments.forEach((d) => this.discoveredDocuments.add(d));
        },
        onEnd: async (result, importStubs) => {
          this.options.processor.pushResult("document", result, importStubs);
          await this.options.processor.process();
          this.building = false;
        },
      });

      void this.docsCtx.watch();

      await oldDocsCtx?.dispose();
    }
    setImmediate(() => {
      if (this.running) void this.tick();
    });
  }

  async finish(): Promise<void> {
    while (this.building) {
      await new Promise(setImmediate);
    }
    await Promise.all([this.docsCtx?.dispose(), this.worldCtx?.dispose()]);
    this.running = false;
  }
}
