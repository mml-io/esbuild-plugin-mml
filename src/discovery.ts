import * as esbuild from "esbuild";
import * as path from "path";
import * as fs from "fs";
import { DiscoveredEntryPoints } from "./types";

export async function doDiscoveryBuild(
  esbuildOptions: esbuild.BuildOptions,
): Promise<DiscoveredEntryPoints> {
  const discovered: DiscoveredEntryPoints = {
    world: new Set<string>(),
    tsx: new Set<string>(),
    html: new Set<string>(),
  };

  await esbuild.build({
    ...esbuildOptions,
    write: false,
    metafile: true,
    bundle: true,
    plugins: [
      {
        name: "mml-discovery",
        setup(build) {
          build.onResolve({ filter: /.*/ }, async (args) => {
            if (args.pluginData === "mml-skip") return null;

            const isMMLImport = args.path.startsWith("mml:");
            const importPath = isMMLImport ? args.path.slice(4) : args.path;

            if (!isMMLImport && args.kind !== "entry-point") return null;

            const resolveDir =
              args.resolveDir || path.dirname(args.importer || process.cwd());
            let resolvedPath;

            if (args.kind === "entry-point") {
              const absolutePath = path.resolve(resolveDir, importPath);
              if (fs.existsSync(absolutePath)) {
                resolvedPath = { errors: [], path: absolutePath };
              } else {
                resolvedPath = await build.resolve(importPath, {
                  resolveDir,
                  kind: args.kind,
                  pluginData: "mml-skip",
                });
              }
            } else {
              resolvedPath = await build.resolve(importPath, {
                resolveDir,
                kind: args.kind,
                pluginData: "mml-skip",
              });
            }

            if (resolvedPath.errors.length > 0) {
              return isMMLImport
                ? {
                    errors: [
                      {
                        text: `Failed to resolve mml:${importPath} from ${args.importer}`,
                        location: null,
                      },
                    ],
                  }
                : null;
            }

            const isTypeScript = /\.(tsx?|ts)$/.test(resolvedPath.path);
            const isHTML = resolvedPath.path.endsWith(".html");

            if (args.kind === "entry-point") {
              if (isTypeScript) {
                const targetSet = isMMLImport
                  ? discovered.tsx
                  : discovered.world;
                targetSet.add(resolvedPath.path);
                return isMMLImport
                  ? {
                      path: resolvedPath.path,
                      namespace: "mml-discovery-tsx",
                      pluginData: "entry",
                    }
                  : null;
              } else if (isHTML) {
                discovered.html.add(resolvedPath.path);
                return {
                  path: resolvedPath.path,
                  namespace: "mml-discovery-html",
                  pluginData: "entry",
                };
              }
            } else if (isMMLImport) {
              if (isTypeScript) {
                discovered.tsx.add(resolvedPath.path);
                return {
                  path: resolvedPath.path,
                  namespace: "mml-discovery-tsx",
                  pluginData: "import",
                };
              } else if (isHTML) {
                discovered.html.add(resolvedPath.path);
                return {
                  path: resolvedPath.path,
                  namespace: "mml-discovery-html",
                  pluginData: "import",
                };
              }
            }

            return null;
          });

          build.onLoad(
            { filter: /.*/, namespace: "mml-discovery-tsx" },
            async (args) => {
              const content = await fs.promises.readFile(args.path, "utf-8");

              if (args.pluginData === "entry") {
                const withExport = content.includes("export default")
                  ? content
                  : `${content}\nexport default "";`;
                return {
                  contents: withExport,
                  loader: "tsx",
                  resolveDir: path.dirname(args.path),
                };
              }

              // For imports, we need to process the file but also ensure it has a default export
              const withExport = content.includes("export default")
                ? content
                : `${content}\nexport default "";`;
              return {
                contents: withExport,
                loader: "tsx",
                resolveDir: path.dirname(args.path),
              };
            },
          );

          build.onLoad(
            { filter: /.*/, namespace: "mml-discovery-html" },
            async (args) => {
              if (args.pluginData === "import") {
                return { contents: '""', loader: "text" };
              }

              const content = await fs.promises.readFile(args.path, "utf-8");
              return {
                contents: content,
                loader: "text",
                resolveDir: path.dirname(args.path),
              };
            },
          );
        },
      },
    ],
  });

  return discovered;
}
