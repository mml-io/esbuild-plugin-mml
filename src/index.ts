export * from "./plugin";
export * from "./world-config";
