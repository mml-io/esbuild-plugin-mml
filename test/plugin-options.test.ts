import * as esbuild from "esbuild";
import { mml } from "../src";
import fsp from "node:fs/promises";
import path from "node:path";
import { waitForDispose, walk } from "./test-helpers";

describe("Plugin Options and Configuration", () => {
  const testOutdir = "test/generated/plugin-options";

  beforeEach(async () => {
    await fsp.rm(testOutdir, { recursive: true, force: true });
  });

  it("should respect verbose option for enhanced logging", async () => {
    const outdir = path.join(testOutdir, "verbose-true");
    await fsp.rm(outdir, { recursive: true, force: true });
    
    // Capture console output
    const originalLog = console.log;
    const logs: string[] = [];
    console.log = (...args) => {
      logs.push(args.join(' '));
      originalLog(...args);
    };

    try {
      const config = {
        outdir,
        entryPoints: ["mml:test/src/a.ts"],
        loader: {
          ".glb": "file" as const,
        },
        plugins: [mml({
          verbose: true,
          assetDir: "assets",
        })],
      };

      const result = await esbuild.build(config);

      // Check for build errors
      if (result.errors.length > 0) {
        throw new Error(`Build failed with errors: ${result.errors.map(e => e.text).join(', ')}`);
      }

      // Verify that verbose logging occurred
      expect(logs.length).toBeGreaterThan(0);
      expect(logs.some(log => log.includes('discovered entry points'))).toBe(true);
    } finally {
      console.log = originalLog;
    }
  });

  it("should have minimal logging with verbose disabled", async () => {
    const outdir = path.join(testOutdir, "verbose-false");
    await fsp.rm(outdir, { recursive: true, force: true });
    
    // Capture console output
    const originalLog = console.log;
    const logs: string[] = [];
    console.log = (...args) => {
      logs.push(args.join(' '));
      originalLog(...args);
    };

    try {
      const config = {
        outdir,
        entryPoints: ["mml:test/src/a.ts"],
        loader: {
          ".glb": "file" as const,
        },
        plugins: [mml({
          verbose: false, // Explicitly set to false
          assetDir: "assets",
        })],
      };

      const result = await esbuild.build(config);

      // Check for build errors
      if (result.errors.length > 0) {
        throw new Error(`Build failed with errors: ${result.errors.map(e => e.text).join(', ')}`);
      }

      // With verbose disabled, should not have detailed [mml]: logs
      const mmlDetailedLogs = logs.filter(log => log.includes('[mml]:'));
      expect(mmlDetailedLogs.length).toBe(0);
    } finally {
      console.log = originalLog;
    }
  });

  it("should handle all plugin options together", async () => {
    const outdir = path.join(testOutdir, "all-options");
    await fsp.rm(outdir, { recursive: true, force: true });
    
    const config = {
      outdir,
      entryPoints: ["mml:test/src/a.ts"],
      loader: {
        ".glb": "file" as const,
      },
      plugins: [mml({
        verbose: true,
        documentPrefix: "https://example.com/docs/",
        assetPrefix: "https://cdn.example.com/assets/",
        assetDir: "custom-assets",
      })],
    };

    const result = await esbuild.build(config);

    // Check for build errors
    if (result.errors.length > 0) {
      throw new Error(`Build failed with errors: ${result.errors.map(e => e.text).join(', ')}`);
    }

    await waitForDispose();

    // Check that manifest was created with all specified values
    const manifestPath = path.join(outdir, "manifest.json");
    const manifestContent = await fsp.readFile(manifestPath, "utf-8");
    const manifest = JSON.parse(manifestContent);
    
    expect(manifest.documentPrefix).toBe("https://example.com/docs/");
    expect(manifest.assetPrefix).toBe("https://cdn.example.com/assets/");

    // Verify that assets were copied to the custom directory
    const assetDir = path.join(outdir, "custom-assets");
    const assetExists = await fsp.access(assetDir).then(() => true).catch(() => false);
    expect(assetExists).toBe(true);
  });

  it("should handle edge cases in option values", async () => {
    const outdir = path.join(testOutdir, "edge-cases");
    await fsp.rm(outdir, { recursive: true, force: true });
    
    const config = {
      outdir,
      entryPoints: ["mml:test/src/a.ts"],
      loader: {
        ".glb": "file" as const,
      },
      plugins: [mml({
        documentPrefix: "", // Empty string
        assetPrefix: "//special-chars!@#$/",
        assetDir: "assets",
      })],
    };

    const result = await esbuild.build(config);

    // Check for build errors
    if (result.errors.length > 0) {
      throw new Error(`Build failed with errors: ${result.errors.map(e => e.text).join(', ')}`);
    }

    await waitForDispose();

    // Check that manifest was created with the specified values
    const manifestPath = path.join(outdir, "manifest.json");
    const manifestContent = await fsp.readFile(manifestPath, "utf-8");
    const manifest = JSON.parse(manifestContent);
    
    expect(manifest.documentPrefix).toBe("");
    expect(manifest.assetPrefix).toBe("//special-chars!@#$/");
  });

  it("should use appropriate defaults when options are not provided", async () => {
    const outdir = path.join(testOutdir, "default-options");
    await fsp.rm(outdir, { recursive: true, force: true });
    
    // Test with no options provided
    const config = {
      outdir,
      entryPoints: ["mml:test/src/a.ts"],
      loader: {
        ".glb": "file" as const,
      },
      plugins: [mml()], // No options provided
    };

    const result = await esbuild.build(config);

    // Check for build errors and fail the test if any exist
    if (result.errors.length > 0) {
      throw new Error(`Build failed with errors: ${result.errors.map(e => e.text).join(', ')}`);
    }

    await waitForDispose();

    // Check that manifest was created with default values
    const manifestPath = path.join(outdir, "manifest.json");
    const manifestContent = await fsp.readFile(manifestPath, "utf-8");
    const manifest = JSON.parse(manifestContent);
    
    // Default values should be applied
    expect(manifest.documentPrefix).toBe("ws:///");
    expect(manifest.assetPrefix).toBe("/");
    expect(manifest.assetNameToPath).toBeDefined();

    for await (const { path, content } of walk(outdir)) {
      expect(content).toMatchSnapshot(path);
    }
  });



  it("should strip html extension when stripHtmlExtension is true", async () => {
    const outdir = path.join(testOutdir, "strip-html-extension");
    await fsp.rm(outdir, { recursive: true, force: true });
    
    const config = {
      outdir,
      entryPoints: ["mml:test/src/a.ts"],
      loader: {
        ".glb": "file" as const,
      },
      plugins: [mml({
        stripHtmlExtension: true,
      })],
    };

    const result = await esbuild.build(config);

    if (result.errors.length > 0) {
      throw new Error(`Build failed with errors: ${result.errors.map(e => e.text).join(', ')}`);
    }

    await waitForDispose();

    // Check that manifest was created and URLs don't have .html extensions
    const manifestPath = path.join(outdir, "manifest.json");
    const manifestContent = await fsp.readFile(manifestPath, "utf-8");
    const manifest = JSON.parse(manifestContent);
    
    // The document URLs should not have .html extensions
    for (const docUrl of Object.values(manifest.documentNameToPath)) {
      expect(docUrl).not.toMatch(/\.html$/);
    }

    for await (const { path, content } of walk(outdir)) {
      expect(content).toMatchSnapshot(path);
    }
  });
}); 