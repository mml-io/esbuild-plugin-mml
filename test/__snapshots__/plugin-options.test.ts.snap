// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plugin Options and Configuration should strip html extension when stripHtmlExtension is true: generated/plugin-options/strip-html-extension/a.js 1`] = `
"(() => {
  var b_default = "ws:///test-src-b";
  var d_default = "ws:///test-src-c-d";
  var c = "wat-" + d_default;
  var duck_default = "/test_src_duck.glb";
  console.log(b_default, d_default, duck_default);
})();
"
`;

exports[`Plugin Options and Configuration should strip html extension when stripHtmlExtension is true: generated/plugin-options/strip-html-extension/assets/test_src_duck.glb 1`] = `"test_src_duck.glb"`;

exports[`Plugin Options and Configuration should strip html extension when stripHtmlExtension is true: generated/plugin-options/strip-html-extension/b.js 1`] = `
"(() => {
  var d_default = "ws:///test-src-c-d";
  var c = "wat-" + d_default;
  var b = { c, d: d_default };
  var b_default = b;
})();
"
`;

exports[`Plugin Options and Configuration should strip html extension when stripHtmlExtension is true: generated/plugin-options/strip-html-extension/c/d.js 1`] = `
"(() => {
  var d_default = "<html>\\n  <body>\\n    <p>I'm html!</p>\\n  </body>\\n</html>\\n";
})();
"
`;

exports[`Plugin Options and Configuration should strip html extension when stripHtmlExtension is true: generated/plugin-options/strip-html-extension/manifest.json 1`] = `
"{
  "worlds": [],
  "documentNameToPath": {
    "test-src-a": "test-src-a",
    "test-src-b": "test-src-b",
    "test-src-c-d": "test-src-c-d"
  },
  "assetNameToPath": {
    "test_src_duck.glb": "assets/test_src_duck.glb"
  },
  "documentPrefix": "ws:///",
  "assetPrefix": "/"
}"
`;

exports[`Plugin Options and Configuration should strip html extension when stripHtmlExtension is true: generated/plugin-options/strip-html-extension/test-src-a 1`] = `"test-src-a"`;

exports[`Plugin Options and Configuration should strip html extension when stripHtmlExtension is true: generated/plugin-options/strip-html-extension/test-src-b 1`] = `"test-src-b"`;

exports[`Plugin Options and Configuration should strip html extension when stripHtmlExtension is true: generated/plugin-options/strip-html-extension/test-src-c-d 1`] = `"test-src-c-d"`;

exports[`Plugin Options and Configuration should use appropriate defaults when options are not provided: generated/plugin-options/default-options/a.js 1`] = `
"(() => {
  var b_default = "ws:///test-src-b.html";
  var d_default = "ws:///test-src-c-d.html";
  var c = "wat-" + d_default;
  var duck_default = "/test_src_duck.glb";
  console.log(b_default, d_default, duck_default);
})();
"
`;

exports[`Plugin Options and Configuration should use appropriate defaults when options are not provided: generated/plugin-options/default-options/assets/test_src_duck.glb 1`] = `"test_src_duck.glb"`;

exports[`Plugin Options and Configuration should use appropriate defaults when options are not provided: generated/plugin-options/default-options/b.js 1`] = `
"(() => {
  var d_default = "ws:///test-src-c-d.html";
  var c = "wat-" + d_default;
  var b = { c, d: d_default };
  var b_default = b;
})();
"
`;

exports[`Plugin Options and Configuration should use appropriate defaults when options are not provided: generated/plugin-options/default-options/c/d.js 1`] = `
"(() => {
  var d_default = "<html>\\n  <body>\\n    <p>I'm html!</p>\\n  </body>\\n</html>\\n";
})();
"
`;

exports[`Plugin Options and Configuration should use appropriate defaults when options are not provided: generated/plugin-options/default-options/manifest.json 1`] = `
"{
  "worlds": [],
  "documentNameToPath": {
    "test-src-a.html": "test-src-a.html",
    "test-src-b.html": "test-src-b.html",
    "test-src-c-d.html": "test-src-c-d.html"
  },
  "assetNameToPath": {
    "test_src_duck.glb": "assets/test_src_duck.glb"
  },
  "documentPrefix": "ws:///",
  "assetPrefix": "/"
}"
`;

exports[`Plugin Options and Configuration should use appropriate defaults when options are not provided: generated/plugin-options/default-options/test-src-a.html 1`] = `
"<body></body><script>(() => {
  var b_default = "ws:///test-src-b.html";
  var d_default = "ws:///test-src-c-d.html";
  var c = "wat-" + d_default;
  var duck_default = "/test_src_duck.glb";
  console.log(b_default, d_default, duck_default);
})();
</script>"
`;

exports[`Plugin Options and Configuration should use appropriate defaults when options are not provided: generated/plugin-options/default-options/test-src-b.html 1`] = `
"<body></body><script>(() => {
  var d_default = "ws:///test-src-c-d.html";
  var c = "wat-" + d_default;
  var b = { c, d: d_default };
  var b_default = b;
})();
</script>"
`;

exports[`Plugin Options and Configuration should use appropriate defaults when options are not provided: generated/plugin-options/default-options/test-src-c-d.html 1`] = `
"<html>
  <body>
    <p>I'm html!</p>
  </body>
</html>
"
`;
